In [1]:
import pandas as pd
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.metrics import pairwise_distances

Idea: similarity is decided by how much percentage of their budget is used up.

| UserID | TipID1 | TipID2 | TipID3 | ... |
|--------|------|------|------|-----|
| thaer  | 0  | 0 | 1 | ... |
| erwoh  | 1 | 1 | 0 | ... |
| wobll  | 1 | 0 | 0 | ... |
| ...    | ...  | ...  | ...  | ... |


In [2]:
data = pd.read_csv('data.csv')
data['userID'] = data.index
data.drop('Unnamed: 0', axis=1, inplace=True)

In [3]:
data.head()

Budget  Education   Food  Health  Housing  Personal  Student  Tip 1  \
0  4077.0    1635.77  568.0   236.0   1033.0     628.0        0    0.0   
1  4101.0    1593.77  519.0   299.0    908.0     573.0        1    1.0   
2  3829.0    1633.77  521.0   286.0    821.0     298.0        2    1.0   
3  3839.0    1670.77  571.0   280.0    990.0     628.0        3    1.0   
4  4208.0    1611.77  307.0   214.0    957.0     365.0        4    1.0   

   Tip 10  Tip 11   ...    Tip 20  Tip 3  Tip 4  Tip 5  Tip 6  Tip 7  Tip 8  \
0     0.0     0.0   ...       1.0    0.0    0.0    1.0    0.0    1.0    1.0   
1     0.0     0.0   ...       1.0    1.0    1.0    1.0    1.0    1.0    1.0   
2     1.0     0.0   ...       0.0    0.0    1.0    0.0    1.0    1.0    0.0   
3     1.0     0.0   ...       1.0    0.0    0.0    1.0    0.0    0.0    0.0   
4     1.0     1.0   ...       1.0    0.0    1.0    0.0    1.0    0.0    1.0   

   Tip 9  Transportation  userID  
0    1.0            82.0       0  
1    0.0            67.0       1  
2    1.0            81.0       2  
3    1.0            69.0       3  
4    1.0            78.0       4  

[5 rows x 29 columns]

In [4]:
Ratings = data.iloc[:, np.arange(7, 29)]
Ratings.drop('Transportation', axis=1, inplace=True)
Ratings.head()

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


Tip 1  Tip 10  Tip 11  Tip 12  Tip 13  Tip 14  Tip 15  Tip 16  Tip 17  \
0    0.0     0.0     0.0     1.0     1.0     0.0     1.0     0.0     0.0   
1    1.0     0.0     0.0     0.0     0.0     0.0     0.0     1.0     0.0   
2    1.0     1.0     0.0     1.0     0.0     1.0     0.0     0.0     0.0   
3    1.0     1.0     0.0     1.0     0.0     1.0     1.0     1.0     1.0   
4    1.0     1.0     1.0     1.0     1.0     0.0     0.0     1.0     0.0   

   Tip 18   ...    Tip 2  Tip 20  Tip 3  Tip 4  Tip 5  Tip 6  Tip 7  Tip 8  \
0     1.0   ...      0.0     1.0    0.0    0.0    1.0    0.0    1.0    1.0   
1     0.0   ...      1.0     1.0    1.0    1.0    1.0    1.0    1.0    1.0   
2     0.0   ...      1.0     0.0    0.0    1.0    0.0    1.0    1.0    0.0   
3     1.0   ...      0.0     1.0    0.0    0.0    1.0    0.0    0.0    0.0   
4     1.0   ...      1.0     1.0    0.0    1.0    0.0    1.0    0.0    1.0   

   Tip 9  userID  
0    1.0       0  
1    0.0       1  
2    1.0       2  
3    1.0       3  
4    1.0       4  

[5 rows x 21 columns]

In [5]:
Ratings_melted = pd.melt(Ratings, id_vars='userID', value_vars=Ratings.columns.values[:20])
Ratings_melted.head()

userID variable  value
0       0    Tip 1    0.0
1       1    Tip 1    1.0
2       2    Tip 1    1.0
3       3    Tip 1    1.0
4       4    Tip 1    1.0

In [6]:
# Replacing NaN Average
Ratings = Ratings.fillna(Ratings.mean(axis=0))
Ratings.head()

Tip 1  Tip 10  Tip 11  Tip 12  Tip 13  Tip 14  Tip 15  Tip 16  Tip 17  \
0    0.0     0.0     0.0     1.0     1.0     0.0     1.0     0.0     0.0   
1    1.0     0.0     0.0     0.0     0.0     0.0     0.0     1.0     0.0   
2    1.0     1.0     0.0     1.0     0.0     1.0     0.0     0.0     0.0   
3    1.0     1.0     0.0     1.0     0.0     1.0     1.0     1.0     1.0   
4    1.0     1.0     1.0     1.0     1.0     0.0     0.0     1.0     0.0   

   Tip 18   ...    Tip 2  Tip 20  Tip 3  Tip 4  Tip 5  Tip 6  Tip 7  Tip 8  \
0     1.0   ...      0.0     1.0    0.0    0.0    1.0    0.0    1.0    1.0   
1     0.0   ...      1.0     1.0    1.0    1.0    1.0    1.0    1.0    1.0   
2     0.0   ...      1.0     0.0    0.0    1.0    0.0    1.0    1.0    0.0   
3     1.0   ...      0.0     1.0    0.0    0.0    1.0    0.0    0.0    0.0   
4     1.0   ...      1.0     1.0    0.0    1.0    0.0    1.0    0.0    1.0   

   Tip 9  userID  
0    1.0       0  
1    0.0       1  
2    1.0       2  
3    1.0       3  
4    1.0       4  

[5 rows x 21 columns]

In [7]:
cosine = cosine_similarity(Ratings.drop('userID', axis=1))
np.fill_diagonal(cosine, 0)
similarity_with_tip = pd.DataFrame(cosine,index=Ratings.index)
similarity_with_tip.columns=Ratings.index
similarity_with_tip

0         1         2         3         4         5         6   \
0   0.000000  0.400000  0.400000  0.639010  0.591608  0.597614  0.421637   
1   0.400000  0.000000  0.500000  0.365148  0.591608  0.358569  0.527046   
2   0.400000  0.500000  0.000000  0.547723  0.676123  0.478091  0.316228   
3   0.639010  0.365148  0.547723  0.000000  0.617213  0.218218  0.673575   
4   0.591608  0.591608  0.676123  0.617213  0.000000  0.606092  0.445435   
5   0.597614  0.358569  0.478091  0.218218  0.606092  0.000000  0.125988   
6   0.421637  0.527046  0.316228  0.673575  0.445435  0.125988  0.000000   
7   0.421637  0.421637  0.421637  0.384900  0.623610  0.503953  0.111111   
8   0.456435  0.639010  0.547723  0.500000  0.694365  0.436436  0.481125   
9   0.632456  0.158114  0.316228  0.288675  0.267261  0.377964  0.000000   
10  0.600000  0.700000  0.500000  0.547723  0.507093  0.478091  0.632456   
11  0.476731  0.572078  0.572078  0.609272  0.483494  0.341882  0.703526   
12  0.821584  0.547723  0.456435  0.583333  0.617213  0.545545  0.481125   
13  0.447214  0.447214  0.559017  0.408248  0.472456  0.267261  0.471405   
14  0.639010  0.547723  0.547723  0.500000  0.771517  0.654654  0.288675   
15  0.476731  0.476731  0.476731  0.435194  0.483494  0.683763  0.301511   
16  0.476731  0.572078  0.572078  0.435194  0.644658  0.455842  0.301511   
17  0.526235  0.613941  0.613941  0.640513  0.592999  0.209657  0.647150   
18  0.572078  0.476731  0.572078  0.696311  0.564076  0.341882  0.603023   
19  0.335410  0.447214  0.559017  0.408248  0.755929  0.534522  0.235702   
20  0.500000  0.500000  0.700000  0.730297  0.676123  0.358569  0.316228   
21  0.421637  0.737865  0.527046  0.384900  0.623610  0.503953  0.444444   
22  0.358569  0.239046  0.478091  0.545545  0.606092  0.428571  0.503953   
23  0.527046  0.421637  0.527046  0.384900  0.623610  0.503953  0.333333   
24  0.447214  0.559017  0.447214  0.510310  0.377964  0.400892  0.471405   
25  0.421637  0.421637  0.421637  0.288675  0.623610  0.377964  0.333333   
26  0.239046  0.239046  0.478091  0.327327  0.505076  0.285714  0.251976   
27  0.547723  0.547723  0.456435  0.583333  0.540062  0.327327  0.673575   
28  0.600000  0.500000  0.200000  0.456435  0.591608  0.478091  0.421637   
29  0.316228  0.527046  0.527046  0.673575  0.623610  0.251976  0.666667   
..       ...       ...       ...       ...       ...       ...       ...   
70  0.387298  0.129099  0.258199  0.353553  0.436436  0.617213  0.272166   
71  0.500000  0.600000  0.500000  0.547723  0.507093  0.358569  0.527046   
72  0.527046  0.210819  0.421637  0.577350  0.534522  0.377964  0.333333   
73  0.447214  0.335410  0.447214  0.612372  0.661438  0.400892  0.353553   
74  0.507093  0.591608  0.591608  0.694365  0.714286  0.404061  0.534522   
75  0.456435  0.639010  0.456435  0.583333  0.617213  0.436436  0.577350   
76  0.632456  0.421637  0.210819  0.481125  0.445435  0.503953  0.666667   
77  0.527046  0.527046  0.527046  0.481125  0.712697  0.377964  0.444444   
78  0.527046  0.527046  0.421637  0.577350  0.534522  0.125988  0.444444   
79  0.762770  0.476731  0.572078  0.522233  0.725241  0.683763  0.402015   
80  0.639010  0.456435  0.456435  0.750000  0.617213  0.327327  0.673575   
81  0.456435  0.639010  0.547723  0.666667  0.694365  0.218218  0.577350   
82  0.572078  0.572078  0.667424  0.609272  0.644658  0.227921  0.502519   
83  0.400000  0.700000  0.300000  0.547723  0.507093  0.239046  0.632456   
84  0.667424  0.572078  0.381385  0.696311  0.725241  0.341882  0.502519   
85  0.667424  0.572078  0.381385  0.522233  0.564076  0.455842  0.502519   
86  0.527046  0.527046  0.316228  0.577350  0.623610  0.251976  0.444444   
87  0.591608  0.676123  0.507093  0.540062  0.642857  0.404061  0.534522   
88  0.424264  0.282843  0.141421  0.387298  0.239046  0.169031  0.596285   
89  0.639010  0.456435  0.547723  0.583333  0.617213  0.436436  0.673575   
90  0.613941  0.701646  0.613941  0.640513  0.667124  0.41

In [8]:
def find_n_neighbours(df,n):
    order = np.argsort(df.values, axis=1)[:, :n]
    df = df.apply(lambda x: pd.Series(x.sort_values(ascending=False)
           .iloc[:n].index, 
          index=['top{}'.format(i) for i in range(1, n+1)]), axis=1)
    return df

In [9]:
# top 5 neighbours for each user
sim_user_5_m = find_n_neighbours(similarity_with_tip,5)
sim_user_5_m

top1  top2  top3  top4  top5
0     12    97    79    36    32
1     64    21    56    92    90
2     93    34    45    62    36
3     49    80    20    33    44
4     52    36    44    45    34
5     15    79    36    64    14
6     58    49    11    56    57
7     61     8    14    59    15
8      7    14    46    56    83
9      0    79    12    14    96
10    97    92    33     1    53
11    49    18     6    56    68
12     0    79    95    14    97
13    32    17    90    50    74
14    46    79     4     7    34
15    92    95    42     7    35
16    52    37    45    87    78
17    38    27    89    90    82
18    55    44    54    79    11
19     4    14    52    77    23
20     3    34    81    41     2
21    92     1    24    43    93
22    44    49    80    14    34
23    52    39    45    38    87
24    91    33    27    57    21
25    87    36     8    14     7
26    42    36    67    25    72
27    17    74    38    95    89
28    85    87    80    94    40
29    66    62    81    44    73
..   ...   ...   ...   ...   ...
70    72     7    48    31     5
71    87    57    17    39    53
72    73    70    81     7    96
73    34    81    29    62    66
74    59    84    61    95    38
75    92    31    48    41    83
76    89    39    56    57     6
77    86     4    19    54    52
78    82    96    87    38    16
79    12    14    36     0    97
80    49    89     3    28    91
81    82    66    29    52    20
82    45    78    81    97    17
83    92     8    75    97    90
84    74    38    90     4    59
85    87    28    47    86    36
86    77    94    84    85    27
87    85    39    28    71    36
88    80    89    27    13    17
89    17    38    76    27    80
90    97    17    32    84    59
91    24    49    65    95    74
92    75    15    64    83    95
93     2    34    33    45     4
94    86    44    32     0    28
95    49    15    59    74    12
96    98    42    78    57    34
97    90     0    32    79    82
98    46    96     4     3    34
99    91    74    84    27    80

[100 rows x 5 columns]

In [10]:
def get_user_similar_tips(user1, user2):
    common_tips = Ratings_melted[Ratings_melted['userID'] == user1].merge(
        Ratings_melted[Ratings_melted.userID == user2],
        on = "variable",
        how = "inner" )
    return common_tips# .merge( tips, on = 'tipID' )

In [11]:
opinion = get_user_similar_tips(37,86)
opinion

userID_x variable  value_x  userID_y  value_y
0         37    Tip 1      0.0        86      1.0
1         37   Tip 10      1.0        86      1.0
2         37   Tip 11      1.0        86      0.0
3         37   Tip 12      0.0        86      0.0
4         37   Tip 13      0.0        86      1.0
5         37   Tip 14      0.0        86      0.0
6         37   Tip 15      1.0        86      1.0
7         37   Tip 16      1.0        86      1.0
8         37   Tip 17      0.0        86      0.0
9         37   Tip 18      0.0        86      1.0
10        37   Tip 19      0.0        86      0.0
11        37    Tip 2      0.0        86      0.0
12        37   Tip 20      0.0        86      0.0
13        37    Tip 3      1.0        86      0.0
14        37    Tip 4      1.0        86      1.0
15        37    Tip 5      0.0        86      1.0
16        37    Tip 6      0.0        86      0.0
17        37    Tip 7      0.0        86      0.0
18        37    Tip 8      0.0        86      1.0
19        37    Tip 9      0.0        86      0.0

In [12]:
import operator

def recommended(user1): 
    sim = sim_user_5_m.loc[user1, :].values.tolist()
    dic = {}
    for u in sim: 
        u_list = Ratings_melted[Ratings_melted['userID'] == u]
        u_list = u_list[u_list['value'] == 1.0]['variable'].unique()
        for t in u_list:
            if t in dic:
                dic[t] += 1
            else:
                dic[t] = 1
    sorted_x = sorted(dic.items(), key=operator.itemgetter(1), reverse=True)
    return [it[0] for it in sorted_x[:3]]

In [13]:
recommended(1)

['Tip 16', 'Tip 20', 'Tip 4']

In [14]:
user = int(input("Enter the user id for whom you want to recommend some tips: "))
for i in recommended(user):
    print(i)

Enter the user id for whom you want to recommend some tips: 19
Tip 18
Tip 19
Tip 2


### References
- https://medium.com/sfu-big-data/recommendation-systems-user-based-collaborative-filtering-using-n-nearest-neighbors-bf7361dc24e0
- https://github.com/ashaypathak/Recommendation-system/blob/master/Movie_Recommendation.ipynb
- https://en.wikipedia.org/wiki/Collaborative_filtering#Memory-based
- https://en.wikipedia.org/wiki/Locality-sensitive_hashing
- https://pypi.org/project/LocalitySensitiveHashing/
- https://www.analyticsvidhya.com/blog/2018/06/comprehensive-guide-recommendation-engine-python/

### Ideas for future implementations

User-based collaborative filtering so that we will suggest tips based on how much of their budget has been spent (to hopefully avoid having similar tips being recommended to one user) <- also not sure that that is an issue yet. 
- Idea: similarity is decided by how much percentage of their budget is used up.

| UserID | Cat1 | Cat2 | Cat3 | ... |
|--------|------|------|------|-----|
| thaer  | .734 | .238 | 1.32 | ... |
| erwoh  | .781 | .362 | .987 | ... |
| wobll  | .294 | .272 | .549 | ... |
| ...    | ...  | ...  | ...  | ... |